### ISSUE TITLE
Azure Storage

### ISSUE THREAD

**Issue by SCKelemen** [upspin/upspin#531]
Is there an Azure implementation for Storage, yet? I'll be at a hackathon this weekend, and this is something I could probably accomplish as a project, using the gcp, aws, backblaze, drive, and dropbox implementations as a reference.

**Comment by gbbr** [comment]
The official Google Drive repo is not yet created (#479) but if you need inspiration you can find it on my account: [gbbr/drive.upspin.io](https://github.com/gbbr/drive.upspin.io). But any one of those examples should be clear and easy enough anyway!

Have fun at the hackathon!

**Comment by SCKelemen** [comment]
@gbbr https://github.com/gbbr/drive.upspin.io/blob/master/config/oauth2.go
I'm not sure if you want to keep clientid and secrets in your repo.

**Comment by gbbr** [comment]
Your job title says "Security Engineer" so I've safely assumed you aren't a malicious one 😊 

But on a more serious note, I did think of finding solutions to that myself, and initially had a `TODO` there. I couldn't think of a good one for an open-source project, especially after having seen the same in the official [upspin/dropox](https://github.com/upspin/dropbox/) implementation in [setupstorage.go#L81-L88](https://github.com/upspin/dropbox/blob/40ebbecb23f0bc5190aba0e4c129f994f0bd736e/cmd/upspin-setupstorage-dropbox/setupstorage.go#L81-L88).

I'd actually be really happy if you would be able to suggest an idea.

**Comment by SCKelemen** [comment]
I think the best solution I can think of on the top of my head with little thought put into it would be to use an environment variable. Environment variables are pretty easy to setup, and several CI/CD/Container systems support secrets, and can expose them via environment variables. All you have to do then, it check the environment. Go has good support for this. I think it would be good practice to still support configuration code.

**Comment by gbbr** [comment]
They are the Dropbox / Google Drive API developer access tokens for Upspin. They are mandatory for all clients in order to access other accounts on the network (given permission). If you take it away from there, people downloading the Dropbox, Drive, etc. repositories would not be able to set up. The only way I can think of is to ask the users to download their own developer access tokens but that's quite tedious.

**Comment by SCKelemen** [comment]
Yeah, I just want to know if that workflow make sense. Are these required for basic users, they will be joining the network using completely other services, or is this something that only technical people who want to deploy their own instance would need? If it makes sense that technical users would deploy their own, I think its okay to ask them to get their own tokens. Otherwise, its more complicated. Maybe we can create a server that someone would register with. I'd be happy to discuss this further, but for me to truly be able to design a secure system, I would need to know how these pieces fit into the overall architecture of upspin.
